core/src/main/java/io/opencensus/tags/Tag.java

```diff
@@ -0,0 +1,205 @@
+   * @param longFunction the function to call when the tag has a {@code long} value.
+   * @param booleanFunction the function to call when the tag has a {@code boolean} value.
+   * @param <T> The result type of the function.
+   * @return The result of calling the function that matches the tag's type.
+   */
+  public <T> T match(
+      Function<? super TagString, T> stringFunction,
+      Function<? super TagLong, T> longFunction,
+      Function<? super TagBoolean, T> booleanFunction) {
+    return matchWithDefault(
+        stringFunction, longFunction, booleanFunction, TagUtils.<T>throwAssertionError());
+  }
+
+  /**
+   * Applies a function to the tag's key and value. This method is like {@link #match(Function,
+   * Function, Function)}, except that it has a default case, for backwards compatibility when tag
+   * types are added. For example, this code serializes a {@code Tag} and tries to handle new tag
+   * types by calling {@code toString()}.
+   *
+   * <pre>{@code
+   * byte[] serializedValue =
+   *     tag.matchWithDefault(
+   *         stringTag -> serializeString(stringTag.getValue().asString()),
+   *         longTag -> serializeLong(longTag.getValue()),
+   *         booleanTag -> serializeBoolean(booleanTag.getValue()),
+   *         unknownTag -> serializeString(unknownTag.toString()));
+   * }</pre>
+   *
+   * @param stringFunction the function to call when the tag has a {@code String} value.
+   * @param longFunction the function to call when the tag has a {@code long} value.
+   * @param booleanFunction the function to call when the tag has a {@code boolean} value.
+   * @param defaultFunction the function to call when the tag has a value other than {@code String},
+   *     {@code long}, or {@code boolean}.
+   * @param <T> The result type of the function.
+   * @return The result of calling the function that matches the tag's type.
+   */
+  // TODO(sebright): How should we deal with the possibility of adding more tag types in the future?
+  //                 Will this method work? What type of parameter should we use for the default
+  //                 case? Should we remove the non-backwards compatible "match" method?
+  public abstract <T> T matchWithDefault(
+      Function<? super TagString, T> stringFunction,
+      Function<? super TagLong, T> longFunction,
+      Function<? super TagBoolean, T> booleanFunction,
+      Function<? super Tag, T> defaultFunction);
+
+  /** A tag with a {@code String} key and value. */
+  @Immutable
+  @AutoValue
+  public abstract static class TagString extends Tag {
+    TagString() {}
+
+    /**
+     * Creates a {@code TagString} from the given {@code String} key and value.
+     *
+     * @param key the tag key.
+     * @param value the tag value.
+     * @return a {@code TagString} with the given key and value.
+     */
+    public static Tag create(TagKeyString key, TagValueString value) {
```

Comment:
If you return Tag here:
1) what is the point of having TagString public?
2) Can we move the factory methods into Tag? createStringTag, etc.?

Same questions for all instances.